### PR TITLE
chore: Upgrades `teams` resource to auto-generated SDK

### PR DESCRIPTION
**Pull request by oarbusi**
## Description

Upgrades `teams` resource to auto-generated SDK.

Migration tests created.

Link to any related issue(s): CLOUDP-226093

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/common/conversion/misc.go [hunk 1]
```diff
@@ -45,3 +45,11 @@ func ExpandStringListFromSetSchema(set *schema.Set) []string {
 	}
 	return res
 }
+
+func ExpandStringPointerListFromSetSchema(set *schema.Set) *[]string {
```
Comment (wtrocki) [inline internal/common/conversion/misc.go]: [optional] misc.go seems like good candidate for sdk-go class.

File: internal/service/teams/resource_teams.go [hunk 2]
```diff
@@ -208,32 +209,31 @@ func resourceMongoDBAtlasTeamUpdate(ctx context.Context, d *schema.ResourceData,
...
 	id := ids["id"]
 
 	err := retry.RetryContext(ctx, 1*time.Hour, func() *retry.RetryError {
-		_, err := conn.Teams.RemoveTeamFromOrganization(ctx, orgID, id)
+		_, _, err := connV2.TeamsApi.DeleteTeam(ctx, orgID, id).Execute()
 		if err != nil {
-			var target *matlas.ErrorResponse
-			if errors.As(err, &target) && target.ErrorCode == "CANNOT_DELETE_TEAM_ASSIGNED_TO_PROJECT" {
-				projectID, err := getProjectIDByTeamID(ctx, conn, id)
+			if errors.Is(err, errors.New("CANNOT_DELETE_TEAM_ASSIGNED_TO_PROJECT")) {
```
Comment (wtrocki) [inline internal/service/teams/resource_teams.go]: Note: v2 SDK has different way than v1  for checking errors: https://github.com/mongodb/atlas-sdk-go/blob/main/docs/doc_2_error_handling.md

**Review by marcosuma — APPROVED** [review]

File: internal/service/teams/resource_teams.go (hunk 2 above)
Comment (oarbusi) [reply internal/service/teams/resource_teams.go]: Thank! Changed it

File: internal/common/conversion/misc.go (hunk 1 above)
Comment (lantoli) [inline internal/common/conversion/misc.go]: not sure it's worth it to add this function when ExpandStringListFromSetSchema is available.
you can just say result := ExpandStringListFromSetSchema(s) and use &result

File: internal/service/teams/resource_teams.go [hunk 3]
```diff
@@ -61,36 +60,36 @@ func Resource() *schema.Resource {
...
 
 	// Creating the team
-	teamsResp, _, err := conn.Teams.Create(ctx, orgID,
-		&matlas.Team{
+	teamsResp, _, err := connV2.TeamsApi.CreateTeam(ctx, orgID,
+		&admin.Team{
 			Name:      d.Get("name").(string),
-			Usernames: conversion.ExpandStringListFromSetSchema(d.Get("usernames").(*schema.Set)),
-		})
+			Usernames: conversion.ExpandStringPointerListFromSetSchema(d.Get("usernames").(*schema.Set)),
```
Comment (lantoli) [inline internal/service/teams/resource_teams.go]: as above you can do before: 
usernames :=  conversion.ExpandStringPointerListFromSetSchema(d.Get("usernames").(*schema.Set))
and use &usernames here

File: internal/service/teams/resource_teams.go [hunk 4]
```diff
@@ -61,36 +60,36 @@ func Resource() *schema.Resource {
 	}
 }
 
-func resourceMongoDBAtlasTeamCreate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
-	conn := meta.(*config.MongoDBClient).Atlas
+func resourceCreate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	connV2 := meta.(*config.MongoDBClient).AtlasV2
 	orgID := d.Get("org_id").(string)
 
 	// Creating the team
```
Comment (lantoli) [inline internal/service/teams/resource_teams.go]: not sure these comments add value

File: internal/common/conversion/misc.go (hunk 1 above)
Comment (oarbusi) [reply internal/common/conversion/misc.go]: good point! Changing it

File: internal/service/teams/resource_teams.go (hunk 3 above)
Comment (oarbusi) [reply internal/service/teams/resource_teams.go]: done

File: internal/service/teams/resource_teams_test.go [hunk 5]
```diff
@@ -32,42 +32,39 @@ func TestAccConfigRSTeam_basic(t *testing.T) {
 		CheckDestroy:             acc.CheckDestroyTeam,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccMongoDBAtlasTeamConfig(orgID, name,
+				Config: configBasic(orgID, name,
```
Comment (lantoli) [inline internal/service/teams/resource_teams_test.go]: you can change username to this and use it in all places:

username     =[]string{  os.Getenv("MONGODB_ATLAS_USERNAME") }

File: internal/service/teams/resource_teams.go (hunk 4 above)
Comment (oarbusi) [reply internal/service/teams/resource_teams.go]: removed comments

File: internal/service/teams/resource_teams_test.go [hunk 6]
```diff
@@ -13,12 +13,12 @@ import (
...
 	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
 	"github.com/mongodb/terraform-provider-mongodbatlas/internal/testutil/acc"
-	matlas "go.mongodb.org/atlas/mongodbatlas"
+	"go.mongodb.org/atlas-sdk/v20231115005/admin"
 )
 
 func TestAccConfigRSTeam_basic(t *testing.T) {
 	var (
-		team         matlas.Team
+		team         admin.Team
```
Comment (lantoli) [inline internal/service/teams/resource_teams_test.go]: MONGODB_ATLAS_USERNAME variable is not checked in  acc.PreCheckBasic, I reccomend to create a new precheck function cheking that variable as well

File: internal/service/teams/resource_teams_test.go (hunk 5 above)
Comment (oarbusi) [reply internal/service/teams/resource_teams_test.go]: done! Thanks

File: internal/service/teams/resource_teams_test.go [hunk 7]
```diff
@@ -105,15 +102,15 @@ func TestAccConfigRSTeam_importBasic(t *testing.T) {
...
+				ImportStateIdFunc: importStateIDFunc(resourceName),
 				ImportState:       true,
 				ImportStateVerify: true,
 			},
 		},
 	})
 }
 
-func testAccCheckMongoDBAtlasTeamExists(resourceName string, team *matlas.Team) resource.TestCheckFunc {
+func checkExists(resourceName string, team *admin.Team) resource.TestCheckFunc {
```
Comment (lantoli) [inline internal/service/teams/resource_teams_test.go]: i don't think team param is needed, i've changed in other resources to have checkExists only with resourceName and use normal resource.TestCheckResourceAttr for checking all attributes we want, e.g: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1897/files#diff-ffad5b36ece89e3e5aec3c355741582e635a69feae36b8ff4538ff08a2f04f6aR91

**Review by lantoli — APPROVED** [review]

File: internal/service/teams/resource_teams_test.go (hunk 6 above)
Comment (lantoli) [reply internal/service/teams/resource_teams_test.go]: similarly for mig precheck

Comment (oarbusi) [reply internal/service/teams/resource_teams_test.go]: good point! Done

File: internal/testutil/mig/pre_check.go [hunk 8]
```diff
@@ -22,3 +22,9 @@ func PreCheckBasicOwnerID(tb testing.TB) {
 	tb.Helper()
 	PreCheckBasic(tb)
 }
+
+func PreCheckAtlasUsername(tb testing.TB) {
+	tb.Helper()
+	checkAtlasUser(tb)
+	PreCheckBasic(tb)
+}
```
Comment (lantoli) [inline internal/testutil/mig/pre_check.go]: the usual way to implement mig precheck is to call the acc one, e.g.: 
func PreCheckAtlasUsername(tb testing.TB) {
	tb.Helper()
	checkLastVersion(tb)
	acc. PreCheckAtlasUsername(tb)
}

they do the same as acc checks apart from checking last version env.var.

and you can delete checkAtlasUser in mig

File: internal/testutil/mig/provider.go [hunk 9]
```diff
@@ -40,3 +40,10 @@ func checkLastVersion(tb testing.TB) {
 func versionConstraint() string {
 	return os.Getenv("MONGODB_ATLAS_LAST_VERSION")
 }
+
+func checkAtlasUser(tb testing.TB) {
```
Comment (lantoli) [inline internal/testutil/mig/provider.go]: this logic doesn't belong to mig but to acc package

File: internal/testutil/mig/pre_check.go (hunk 8 above)
Comment (oarbusi) [reply internal/testutil/mig/pre_check.go]: Thanks! Done

**Pull request merged**